File: ai/lab3/bfs.py

from collections import deque
# Default Graph
graph = {
    'A': ['B', 'C'],
    'B': ['D', 'E'],
    'C': ['F'],
    'D': [],
    'E': ['F'],
    'F': []
}

#print function
def printPath(path):
    for i in range(len(path)):
        print(path[i], end=",")
    print()

# find paths function
def findPaths(graph, start, end):
    q = deque()    
    path = [start] 
    q.append([start])
  
    while q:
        path = q.popleft()
        last = path[-1]
        if last == end:  
            printPath(path)
        for neighbor in graph[last]:
            if neighbor not in path:
                newPath = path.copy()
                newPath.append(neighbor)
                q.append(newPath)
# call function
findPaths(graph, "A", "F")

# connecting check function
def areConnected(graph, start, end):
    visited = set()
    q = deque()
    q.append(start)
    visited.add(start)

    while q:
        current = q.popleft()

        if current == end:
            return True 

        for neighbor in graph[current]:
            if neighbor not in visited:
                q.append(neighbor)
                visited.add(neighbor)

    return False 
# call function
print(areConnected(graph, "A", "F"))

File: ai/lab3/bfs.js
class Graph {
  constructor() {
    this.adjList = new Map();
  }

  addVertex(vertex) {
    this.adjList.set(vertex, []);
  }

  addEdge(vertex, neighbor) {
    this.adjList.get(vertex).push(neighbor);
    this.adjList.get(neighbor).push(vertex); // Assuming an undirected graph
  }

  bfs(start, destination) {
    // Create a queue for BFS and a visited array to keep track of visited vertices
    const queue = [];
    const visited = new Set();

    // Enqueue the starting vertex
    queue.push([start]);

    // Mark the starting vertex as visited
    visited.add(start);

    // BFS algorithm
    while (queue.length > 0) {
      const path = queue.shift();
      const currentVertex = path[path.length - 1];

      // Check if we reached the destination
      if (currentVertex === destination) {
        return path;
      }

      // Enqueue neighboring vertices
      const neighbors = this.adjList.get(currentVertex);
      for (const neighbor of neighbors) {
        if (!visited.has(neighbor)) {
          const newPath = [...path, neighbor];
          queue.push(newPath);
          visited.add(neighbor);
        }
      }
    }

    // If no path is found
    return null;
  }
}

// Example usage
const graph = new Graph();
graph.addVertex('A');
graph.addVertex('B');
graph.addVertex('C');
graph.addVertex('D');
graph.addEdge('A', 'B');
graph.addEdge('B', 'C');
graph.addEdge('C', 'D');

const path = graph.bfs('A', 'D');
console.log(path);